src/math/power.c

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+ /* Include Files */
+#include <sof/math/power.h>
+#include <sof/debug/panic.h>
+/* Function Definitions */
+/**
+ * Arguments	: int64_t denominator
+ *
+ * Return Type	: int32_t
+ */
+static inline int32_t div_s32s64(int64_t denominator)
+{
+	uint64_t b_denominator;
+	int32_t quotient;
+
+	if (denominator == 0) {
+		assert(denominator);
+	} else {
+		b_denominator =
+		    (denominator < 0) ? ~(uint64_t)denominator + 1 : (uint64_t)denominator;
+		b_denominator = POWER_B_NUMERATOR / b_denominator;
+
+		if (denominator < 0)
+			/* for negative denominator, typecast uint64_t to int32_t for */
+			/* negative quotient */
+			quotient = (int32_t)-(int64_t)b_denominator;
+		else
+			quotient = (int32_t)b_denominator;
+	}
+	return quotient;
+}
+
+/**
+ * Arguments	: int32_t b
+ * base input range [-32 to +32]
+ *
+ *		: int32_t e
+ * exponent input range [ -3 to +3 ]
+ *
+ * Return Type	: int32_t
+ * power output range [-32768 to + 32768]
+ *
+ * +------------------+------------------+------------------+--------+--------+--------+
+ * | base	      | exponent	 | power(returntype)|	b    |	e     |	 p     |
+ * +----+-----+-------+----+-----+-------+----+----+--------+--------+--------+--------+
+ * |WLen| FLen|Signbit|WLen| FLen|Signbit|WLen|FLen|Signbit | Qformat| Qformat| Qformat|
+ * +----+-----+-------+----+-----+-------+----+----+--------+--------+--------+--------+
+ * | 32 | 25  |	 1    | 32 | 29	 |  1	 | 32 | 15 |   1    | 6.26   | 2.30   | 16.16  |
+ * +------------------+------------------+------------------+--------+--------+--------+
+ */
+int32_t power_int32(int32_t b, int32_t e)
+{
+	int32_t i;
+	int32_t k;
+	int32_t multiplier;
+	int32_t p;
+
+	p = POWER_MAX_LIMIT;
+
+	if (e < 0) {
+		e = -e;
+		multiplier = (b == 0) ? INT32_MAX : div_s32s64((int64_t)b);
+	} else {
+		multiplier = b;
+	}
+	i = e >> 29;
+	for (k = 0; k < i; k++)
+		p = (int32_t)((p * (int64_t)multiplier) >> 25);
```

Comment:
This is the hot spot, but just wondering if rounding would improve accuracy here (available in intrinsics shift instruction)?

Reply:
I see `AE_SRAI64` instruction, 
_pseudo code_ 
`p = (int32_t)(AE_SRAI64((int64_t)(multiplier) * (int64_t)(p), (int64_t)(25)));`

Reply:
Did you try rounding in C code if it improves accuracy? Like `((product >> 24) + 1) >> 1`.

Reply:
Thanks . Please see impact of rounding on overall Total harmonic distortion and overall  calculation. 

In the cmocka test `scalar_power.c` file , below information is added. 
/* 'Error (max = 0.000034912111005), THD+N  = -89.1404782100538' */

```

|MINROUND	| MINWOROUND	| MAXROUND	|MAXWOROUND	|THD_WROUND	|THD_WO_ROUND	|
+---------------+---------------+---------------+---------------+---------------+---------------+
|-0.0000150363	|-0.0000043945	|0.0000150363	|0.0000349121	|-96.4572	|-89.1405	|

```


![image](https://user-images.githubusercontent.com/61411947/142853641-c4790c72-42a3-406a-b422-34a744f3303e.png)


![Power_studyround](https://user-images.githubusercontent.com/61411947/142854105-7d4004af-3c0f-4608-85c1-646f84569535.jpg)

Reply:
Thanks @singalsu . Glad that you did rock solid review.